packages/@lwc/template-compiler/src/parser/expression-complex/validate.ts

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { ParserDiagnostics, invariant } from '@lwc/errors';
+import { walk } from 'estree-walker';
+import * as t from '../../shared/estree';
+import type { BaseNode } from 'estree';
+
+const ALWAYS_INVALID_TYPES = new Map(
+    Object.entries({
+        AwaitExpression: 'await expressions',
+        ClassExpression: 'classes',
+        SequenceExpression: 'comma operators',
+        FunctionExpression: 'function expressions',
+        ImportExpression: 'imports',
+        MetaProperty: 'import.meta',
+        NewExpression: 'object instantiation',
+        RegExpLiteral: 'regular expression literals',
+        Super: '`super`',
+        ThisExpression: '`this`',
+        YieldExpression: '`yield`',
+    })
+);
+const STATEMENT_TYPES = new Set([
+    'Statement',
+    'BlockStatement',
+    'BreakStatement',
+    'ContinueStatement',
+    'DebuggerStatement',
+    'DoWhileStatement',
+    'EmptyStatement',
+    'ExpressionStatement',
+    'ForInStatement',
+    'ForStatement',
+    'FunctionDeclaration',
+    'IfStatement',
+    'LabeledStatement',
+    'ReturnStatement',
+    'SwitchStatement',
+    'ThrowStatement',
+    'TryStatement',
+    'VariableDeclaration',
+    'WhileStatement',
+    'WithStatement',
+    'ClassDeclaration',
+    'ExportAllDeclaration',
+    'ExportDefaultDeclaration',
+    'ExportNamedDeclaration',
+    'ForOfStatement',
+    'ImportDeclaration',
+    'DeclareClass',
+    'DeclareFunction',
+    'DeclareInterface',
+    'DeclareModule',
+    'DeclareModuleExports',
+    'DeclareTypeAlias',
+    'DeclareOpaqueType',
+    'DeclareVariable',
+    'DeclareExportDeclaration',
+    'DeclareExportAllDeclaration',
+]);
+const EARLY_STAGE_LANGUAGE_FEATURES = new Set([
```

Comment:
I thought acorn only implemented stage 4 features. Is it possible for acorn to parse expressions with non standard syntax?

Reply:
This is an artifact from my experiments using Babel's parser. You're right, these node types will never be encountered. I'll remove this.